agent/agent.mbt

```diff
@@ -117,19 +123,52 @@ async fn Agent::execute_tool(
         tool_call_id=tool_call.id,
       )
   }
-  let result = tool.call(arguments)
-  let rendered = (tool.render)(result)
-  agent.emit(PostToolCall(tool_call, result~, rendered~))
-  @openai.tool_message(content=rendered, tool_call_id=tool_call.id)
+  let (json, text) = (tool.call)(arguments)
+  agent.emit(PostToolCall(tool_call, result=json, rendered=text))
+  @openai.tool_message(content=text, tool_call_id=tool_call.id)
+}
+
+///|
+trait ToolObject {
+  add_tool(self : Self, agent : Agent) -> Unit noraise
+}
+
+///|
+pub impl[T : ToJson + Show] ToolObject for @tool.Tool[T] with add_tool(
+  self : @tool.Tool[T],
+  agent : Agent,
+) -> Unit noraise {
+  agent.add_tool(self)
 }
 
 ///|
 pub fn Agent::add_tools(
   agent : Agent,
-  tools : Array[@tool.Tool],
+  tools : Array[&ToolObject],
 ) -> Unit noraise {
   for tool in tools {
-    agent.tools[tool.desc.name] = tool
+    tool.add_tool(agent)
+  }
+}
```

Comment:
/cc @bobzhang If we are adding generics to `Tool`, then it is impossible to write something like `Array[forall T . Tool[T]]`, without introducing trait object tricks showing here.